### PR TITLE
feat: collection name everywhere

### PR DESCRIPTION
**Pull request by jonasiwnl**
## Description

This change allows the collection of the vectordb to be switched with the `set_collection` method on the Embedchain class. It also allows the initial collection to be specified using InitConfig, e.g. `InitConfig(collection_name="collection1")`. No new dependencies are required as this is a built-in feature of Chroma.

Fixes # (issue)

Issue 261

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update

## How Has This Been Tested?

This feature has been tested through 2 new tests added to `tests/vectordb/test_chroma_db.py`. It tests that the default collection name is correct (if no name is specified), a custom initial collection name, and switching the collection once the app has already been initialized.

- [x] TestChromaDbDefaultCollection
- [x] TestChromaDbSetCollection

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [x] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

### CODE REVIEW

**Comment by cachho** [comment]
Thanks for the PR, I especially appreciate the inclusion of unit tests. In #305, the configs were reworked. Can you resolve that merge conflict by using the new AppConfig, BaseAppConfig and OpenSourceAppConfig. Thank you!

**Comment by jonasiwnl** [comment]
Sure, I'll get to work on that.

**Comment by jonasiwnl** [comment]
Hey, I think there is a missing dependency `sentence_transformers`. I'm getting an error when trying to initalize an opensourceapp that isn't related to my changes. To recreate:
```python
config = OpenSourceAppConfig()
app = OpenSourceApp(config)
```
am I good to add this dependency in?

**Comment by cachho** [comment]
> Hey, I think there is a missing dependency `sentence_transformers`. I'm getting an error when trying to initalize an opensourceapp that isn't related to my changes. I ran `poetry install --with dev` and all that. To recreate:
> 
> ```python
> config = OpenSourceAppConfig()
> app = OpenSourceApp(config)
> ```
> 
> after adding this dep in, all tests pass and the collection feature works as intended.

this is expected behavior, since we support open source models (and will support many more). If we make them a requirement everyone has to download them, even if they're just using OpenAI.

**Comment by jonasiwnl** [comment]
Okay, fixed

File: pyproject.toml [hunk 1]
```diff
@@ -106,9 +106,5 @@ click = "^8.1.3"
 streamlit = ["streamlit"]
```
Comment (cachho) [inline pyproject.toml:108]: still shows a change in here. please revert.

Comment (jonasiwnl) [reply pyproject.toml]: reverted.

**Comment by cachho** [comment]
I did the merge conflicts for you.

File: embedchain/config/apps/CustomAppConfig.py [hunk 2]
```diff
@@ -21,19 +21,20 @@ def __init__(
...
         embedding_fn_model=None,
         db=None,
+        collection_name=None,
         host=None,
         port=None,
         id=None,
         provider: Providers = None,
         model=None,
         open_source_app_config=None,
-    ):
```
Comment (cachho) [inline embedchain/config/apps/CustomAppConfig.py:30]: please close the bracket, sorry must've been a merge mistake I did.

Comment (jonasiwnl) [reply embedchain/config/apps/CustomAppConfig.py]: Thanks for fixing the merge conflicts! And fixed.

**Review by cachho — APPROVED** [review]

File: embedchain/config/apps/BaseAppConfig.py [hunk 3]
```diff
@@ -8,19 +8,21 @@ class BaseAppConfig(BaseConfig):
...
         :param db: Optional. (Vector) database instance to use for embeddings.
+        :param collection_name: Optional. Collection name for the database.
         :param id: Optional. ID of the app. Document metadata will have this id.
         :param host: Optional. Hostname for the database server.
         :param port: Optional. Port for the database server.
         """
         self._setup_logging(log_level)
 
         self.db = db if db else BaseAppConfig.default_db(embedding_fn=embedding_fn, host=host, port=port)
+        self.collection_name = collection_name
```
Comment (cachho) [inline embedchain/config/apps/BaseAppConfig.py:25]: Does this work with none? or should we use `embedchain` as a default here?

**Review by cachho — CHANGES_REQUESTED** [review]

File: embedchain/config/apps/BaseAppConfig.py (hunk 3 above)
Comment (cachho) [reply embedchain/config/apps/BaseAppConfig.py:25]: please add the default here, that way we can make sure that if people use a where filter they have access to the same name, or if we introduce other databases.

Comment (jonasiwnl) [reply embedchain/config/apps/BaseAppConfig.py]: Done.

File: embedchain/vectordb/chroma_db.py [hunk 4]
```diff
@@ -39,9 +39,11 @@ def _get_or_create_db(self):
         """Get or create the database."""
         return self.client
 
-    def _get_or_create_collection(self):
+    def _get_or_create_collection(self, name=None):
```
Comment (cachho) [inline embedchain/vectordb/chroma_db.py:42]: then I think this is redundant, isn't it?

**Review by cachho — CHANGES_REQUESTED** [review]

File: embedchain/vectordb/chroma_db.py (hunk 4 above)
Comment (jonasiwnl) [reply embedchain/vectordb/chroma_db.py:42]: Sure, but then I have to delete line 6 in `base_vector_db.py`.

**Comment by cachho** [comment]
* unified argument order
* fixed docs
* set default name back to `embedchain_store`, otherwise this would break all existing collections
* added tons of unit tests.

**Comment by cachho** [comment]
This has been thoroughly tested and is ready for production. 

Here's my test script. Compared to the unit tests, this uses `dry_run` instead of `count` , but the result is the same, this might just be easier to understand.

```python
from embedchain import App
from embedchain.config import AppConfig

naval_chat_bot = App()
naval_chat_bot.reset()
print("Reset app to run clean test")
naval_chat_bot = App()

# Embed Online Resources
naval_chat_bot.add("web_page", "https://nav.al/feedback")
naval_chat_bot.add("web_page", "https://nav.al/feedback")

# Embed Local Resources
naval_chat_bot.add_local("qna_pair", ("Who is Naval Ravikant?", "Naval Ravikant is an Indian-American entrepreneur and investor."))

naval_chat_bot.query("What unique capacity does Naval argue humans possess when it comes to understanding explanations or concepts?")
# Answer: Naval argues that humans possess the unique capacity to understand explanations or concepts to the maximum extent possible in this physical reality.

# Test: Add them again, it should show that they are already existing.
naval_chat_bot.add("web_page", "https://nav.al/feedback")
naval_chat_bot.add_local("qna_pair", ("Who is Naval Ravikant?", "Naval Ravikant is an Indian-American entrepreneur and investor."))
# Success

# Test: change database. It should allow to add again.
naval_chat_bot.set_collection('another_collection')
naval_chat_bot.add("web_page", "https://nav.al/feedback")
naval_chat_bot.add_local("qna_pair", ("Who is Naval Ravikant?", "Naval Ravikant is an Indian-American entrepreneur and investor."))
# Success

# Test: One collection should not know anything that's in another collection
naval_chat_bot.set_collection('johns_collection')
naval_chat_bot.add('text','My name is John Doe')
naval_chat_bot.set_collection('janes_collection')
r = naval_chat_bot.query("What is my name?", dry_run=True)
print(r)
# Sucess, shows no context.

# Test: Should allow two apps at the same time. We also test setting a collection through the config.
app1 = App(config=AppConfig(collection_name='first_collection'))
app2 = App(config=AppConfig(collection_name='second_collection'))
# Add data to first collection
app1.add("text", "1")
print("App 1:", app1.count())
# Add data to second collection
app2.add("text", "1")
print("App 2:", app2.count())
# App 1 should be unchanged
print("App 1:", app1.count())
```

**Comment by taranjeet** [comment]
@cachho : can we merge this PR?
have seen the code, looks good.
But there is one comment remaining from your side which needs to be reviewed/addressed or closed.
Let me know.

**Comment by cachho** [comment]
> @cachho : can we merge this PR? have seen the code, looks good. But there is one comment remaining from your side which needs to be reviewed/addressed or closed. Let me know.

good to go

**Pull request merged**